### PR TITLE
Added info on the build system to contributing guide

### PR DESCRIPTION
**Pull request by gchp**
I recently wrote a blog post on contributing to the Rust compiler which
gained some interest. It was mentioned in a comment on Reddit that it
would be useful to integrate some of the information from that post to
the official contributing guide.

This is the start of my efforts to integrate what I wrote with the
official guide.

This commit adds information on the build system. It is not a complete
guide on the build system, but it should be enough to provide a good
starting place for those wishing to contribute.

### CODE REVIEW

**Comment by gchp** [comment]
As mentioned in the commit message, this is by no means a full guide to the build system. If there are any pieces you feel should be added to this, please let me know. Hopefully you find it useful!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -77,6 +78,59 @@ to do this is to invoke `rustc` like this:
...
+Rust's setup - there's a lot there, and you can get lost trying to understand
+it all.
+
+If Makefiles are your thing, though, all the configuration lives in
+[the `mk` directory][mkdir] in the project root.
+
+[mkdir]: https://github.com/rust-lang/rust/tree/master/mk/
+
+### Configuration
+Before you can start building the compiler you need to configure the build for
```
Comment (steveklabnik) [inline CONTRIBUTING.md]: could you put a newline above here, like with the other headings?

**Comment by steveklabnik** [comment]
One small nit, this looks good to me, but I'm terrible with the build system, so would like someone else to review too.

**Comment by gchp** [comment]
@steveklabnik updated!

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -77,6 +78,60 @@ to do this is to invoke `rustc` like this:
 $ RUST_BACKTRACE=1 rustc ...
 ```
 
+## The Build System
+
+The build system for Rust is complex. It covers bootstrapping the compiler,
+running tests, building documentation and more. Unless you are familiar with
+Makefiles, I wouldn't suggest trying to understand everything going on in
```
Comment (rkruppe) [inline CONTRIBUTING.md]: * "I" rubs me the wrong way, and above (line 45) "we" is used.
* I wouldn't suggest *anyone* to try and understand everything about the build system, unless they specifically want to improve it (and then, the future-proof way to do that is to improve on #31123 rather than fiddling with the Makefiles).

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -77,6 +78,60 @@ to do this is to invoke `rustc` like this:
...
+```
+./configure
+```
+
+There are large number of options accepted by this script to alter the
+configuration used later in the build process. Some options to note:
+
+- `--enable-debug` - Build a debug version of the compiler (disables optimizations)
+- `--disable-valgrind-rpass` - Don't run tests with valgrind
+- `--enable-clang` - Prefer clang to gcc for building dependencies (ie LLVM)
```
Comment (rkruppe) [inline CONTRIBUTING.md]: Really minor nit but LLVM is not the only C/C++ dependency so this should read "(e.g., LLVM)"

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -77,6 +78,60 @@ to do this is to invoke `rustc` like this:
...
+- `--enable-ccache` - Invoke clang/gcc with ccache to re-use object files between builds
```
Comment (rkruppe) [inline CONTRIBUTING.md]: For most contributors, `--enable-compiler-docs` is invaluable.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -77,6 +78,60 @@ to do this is to invoke `rustc` like this:
...
+root.
+
+```
+./configure
+```
+
+There are large number of options accepted by this script to alter the
+configuration used later in the build process. Some options to note:
+
+- `--enable-debug` - Build a debug version of the compiler (disables optimizations)
```
Comment (rkruppe) [inline CONTRIBUTING.md]: Might want to mention that `--enable-debug --enable-optimized` makes a debug build with optimizations.

File: CONTRIBUTING.md (hunk 2 above)
Comment (gchp) [inline CONTRIBUTING.md]: @rkruppe what about:

> The build system for Rust is complex. It covers bootstrapping the compiler,
> running tests, building documentation and more. It uses some advanced Make
> features which can make it difficult to understand, and one can easily get
> lost in it.

> If Makefiles are your thing, though, all the configuration lives in
> [the `mk` directory][mkdir] in the project root.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -77,6 +78,60 @@ to do this is to invoke `rustc` like this:
...
+
+- `make rustc-stage1` - build up to (and including) the first stage. For most
+  cases we don't need to build the stage2 compiler, so we can save time by not
+  building it. The stage1 compiler is a fully functioning compiler and
+  (probably) will be enough to determine if your change works as expected.
+- `make check` - build the full compiler & run all tests (takes a while). This
+  is what gets run by the continuous integration system against your pull
+  request. You should run this before submitting to make sure your tests pass
+  & everything builds in the correct manner.
+- `make check-stage1-std NO_REBUILD=1` - test the standard library without
```
Comment (rkruppe) [inline CONTRIBUTING.md]: Personally I run `check-stage1-coretest` way more often than `check-stage1-std`. I may be biased because most of my library work ends up in libcore, but then again a huge chunk of the standard library consists of re-exports from core. Also worth mentioning because it deviates from the pattern of all other crates (i.e., `check-stage1-core` does not work).

File: CONTRIBUTING.md (hunk 2 above)
Comment (rkruppe) [inline CONTRIBUTING.md]: IMHO "can easily get lost in it" is more a function of size than advancedness, but otherwise :+1:

Comment (gchp) [inline CONTRIBUTING.md]: What would you suggest as an alternative? Do you think it would be OK to cut that bit off, and leave it as

> ...advanced Make features which can make it difficult to understand.

?

Comment (rkruppe) [inline CONTRIBUTING.md]: Something like:

> [...] building documentation and more.
> 
> If Makefiles are your thing, all the configuration [...] project root. It is easy to get lost though, as the build system is quite large and complex.

sounds more coherent to me.

**Comment by gchp** [comment]
@rkruppe I've updated to reflect your comments. I tweaked the intro paragraph a little more, let me know what you think.

Thank you for your feedback.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -77,6 +78,64 @@ to do this is to invoke `rustc` like this:
...
 
+## The Build System
+
+The build system for Rust is complex. It covers bootstrapping the compiler,
+running tests, building documentation and more.
+
+If Makefiles are your thing, all the configuration lives in
+[the `mk` directory][mkdir] in the project root. Is can be hard to follow
+in places, as it uses some advanced Make features which make for some
+challenging reading. Running `make help` and `make tips` might be useful when
```
Comment (rkruppe) [inline CONTRIBUTING.md]: This document already contains most of the information of `make help`. I would point towards `make nitty-gritty` but I have no idea if it's even remotely up to date. Perhaps cut the false reassurance of documentation and just point people towards `#rust-internals`.

Comment (gchp) [inline CONTRIBUTING.md]: Fair point. Will update now.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -77,6 +78,64 @@ to do this is to invoke `rustc` like this:
 $ RUST_BACKTRACE=1 rustc ...
 ```
 
+## The Build System
+
+The build system for Rust is complex. It covers bootstrapping the compiler,
+running tests, building documentation and more.
```
Comment (rkruppe) [inline CONTRIBUTING.md]: This should have occurred to me sooner, but since this section spends most of its text on how scary the build system is, perhaps the opening paragraph should not mention the complexity at all and instead include some re-assuring words. Something like

> Rust's build system allows you to bootstrap the compiler, run tests and benchmarks, generate documentation, install your freshly-built rust, and more. It's your best friend for compiling and testing your contributions.

Comment (gchp) [inline CONTRIBUTING.md]: Haha, right. Probably a good thing to not just spend time scaring people :)

**Comment by rkruppe** [comment]
Awesome :+1: Thanks for doing this, the build system was by far the part I fought with the most when getting started. I'm still no expert but as far as I am concerned this is good to go.

**Comment by gchp** [comment]
@rkruppe thanks!

Not sure why the build is failing all of a sudden...seems very unlikely that this would cause it to fail?

**Comment by rkruppe** [comment]
I really doubt that this has anything to do with your changes. Travis logs indicate it fails to add a PPA and subsequently can't install some package. Four hours ago, in #31212, the same build commands succeeded so this may be outside of our control.

**Comment by alexcrichton** [comment]
@bors: r+ 05f7b59352abe51f4b1b4897dfcdae7feb33a3a3 rollup

Thanks @gchp!

**Comment by gchp** [comment]
@alexcrichton :+1:
